### ISSUE TITLE
Hibernate ORM SchemaManagementException when running tests

### ISSUE THREAD

**Issue by leonardowestphal** [quarkusio/quarkus#25607]
When updating an application from Quarkus 2.1.4 to version 2.4.0, it started to show a schema validation error when starting the application with the command: mvn quarkus:dev.

```
Failed to start application (with profile dev): org.hibernate.tool.schema.spi.SchemaManagementException: SQL strings added more than once for: '<table_name>'
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.checkExportIdentifier(SchemaCreatorImpl.java:409)
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.createFromMetadata(SchemaCreatorImpl.java:314)
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.performCreation(SchemaCreatorImpl.java:167)
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.doCreation(SchemaCreatorImpl.java:136)
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.doCreation(SchemaCreatorImpl.java:122)
        at org.hibernate.tool.hbm2ddl.SchemaExport.doExecution(SchemaExport.java:314)
        at io.quarkus.hibernate.orm.runtime.devconsole.HibernateOrmDevConsoleInfoSupplier.generateDDL(HibernateOrmDevConsoleInfoSupplier.java:77)
        at io.quarkus.hibernate.orm.runtime.devconsole.HibernateOrmDevConsoleInfoSupplier.pushPersistenceUnit(HibernateOrmDevConsoleInfoSupplier.java:35)
```

I noticed that this error also occurs in versions prior to 2.4.0 and that in version 2.4.1 or higher this error was corrected, however, when running the tests, a similar error occurs: mvn test...
I have the same error with: mvn clean package...

```
java.lang.RuntimeException: java.lang.RuntimeException: Failed to start quarkus
Caused by: java.lang.RuntimeException: Failed to start quarkus
Caused by: java.lang.RuntimeException: javax.persistence.PersistenceException: [PersistenceUnit: Pu1] Unable to build Hibernate SessionFactory
Caused by: javax.persistence.PersistenceException: [PersistenceUnit: Pu1] Unable to build Hibernate SessionFactory
Caused by: org.hibernate.tool.schema.spi.SchemaManagementException: SQL strings added more than once for: <table_name>
```

I also noticed post-boot validation errors:

`(Hibernate post-boot validation thread for App1) Failed to validate Schema: Schema-validation: wrong column type encountered in column...`

As it is a legacy system and I cannot change the hibernate mappings or the database, I ended up using the following properties to resolve post-boot errors:

```
quarkus.hibernate-orm."Pu1".database.generation=none
quarkus.hibernate-orm."Pu1".validate-in-dev-mode=false
quarkus.hibernate-orm."Pu2".database.generation=none
quarkus.hibernate-orm."Pu2".validate-in-dev-mode=false
```

There are two Persistence Units that use the same mappings!
I'm using java 11, Maven 3.8.1, and currently on Quarkus 2.9.0.

@Sanne @yrodiere @gsmet 
Any ideas to solve the schema validation problem when running the tests? I tried to create a new project and reproduce the problem, but I couldn't.

**Comment by leonardowestphal** [comment]
I think I just found the solution.
I believe that some behavior of the test execution profiles has been changed.
Using version 2.1.4, as I was running only locally, I had only set the property: '%dev.quarkus.datasource...'
But with the updated version I also need a property without the 'dev' profile: 'quarkus.datasource...'

The error presented in the log didn't help me much. I think it would be nice to evaluate to see if there is any improvement in the Hibernate ORM extension or another point to make the reason clearer.

**Comment by luisnuxx** [comment]
i'm also facing this issue after upgrading an existing app ( SQL Server ) from Quarkus 1.9.2.Final to 2.9.1.Final

**Comment by yrodiere** [comment]
Thanks for the precisions. A few more questions:

> I solved the problem by adding the "quarkus.datasource" property.

Which one exactly? `quarkus.datasource` is just a prefix.
Also, I can see that you do use `quarkus.datasource` properties in your configuration "without the fix".

> Error stack trace:

Unfortunately, that stack trace is not complete... Any chance you can look into the test report (`target/surefire-reports/something.txt`) and extract the full stack trace from there?

> ```
> Caused by: org.hibernate.tool.schema.spi.SchemaManagementException: SQL strings added more than once for: <schema_name>.<table_name>
> ```

This error most likely happens because two entities are bound to the same table, for some reason. I really wonder how configuring the datasource could solve that problem...

**Comment by leonardowestphal** [comment]
> Which one exactly? `quarkus.datasource` is just a prefix. Also, I can see that you do use `quarkus.datasource` properties in your configuration "without the fix".

quarkus.datasource."DB1".jdbc.url
quarkus.datasource."DB2".jdbc.url

> Unfortunately, that stack trace is not complete... Any chance you can look into the test report (`target/surefire-reports/something.txt`) and extract the full stack trace from there?

The stack trace is the same. I took a look at the .xml file on surefire-reports and found this:

```
WARN  [org.hib.eng.jdb.env.int.JdbcEnvironmentInitiator] (JPA Startup Thread: DB1) HHH000342: Could not obtain connection to query metadata: org.postgresql.util.PSQLException: FATAL: password authentication failed for user "username"
	at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:646)
	at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:180)
	at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:235)
	at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
	at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:223)
	at org.postgresql.Driver.makeConnection(Driver.java:400)
	at org.postgresql.Driver.connect(Driver.java:259)
	at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:226)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:535)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:516)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:75)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1126)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```
If there was an authentication error the application would not even run. And the login data are correct...

> This error most likely happens because two entities are bound to the same table, for some reason. I really wonder how configuring the datasource could solve that problem...

I agree with you. This looks weird but...
I don't have two entities bounded to the same table. If so, the error would occur when running the application too. Maybe it's happening because I use two persistence units with the same mappings/entities (Master and Slave).

**Comment by leonardowestphal** [comment]
I just updated the quarkus version to 2.9.2 and the Java version to 17 and I also changed the name of the PUs (I noticed that one of them had the same name of the database schema and so I decided to change it to test). Now when I run the tests I don't have the error 'SQL strings added more than once for: <schema_name>.<table_name>' anymore.

Now when I run the tests I get this:

```
WARN  [org.hib.eng.jdb.env.int.JdbcEnvironmentInitiator] (JPA Startup Thread: DB1) HHH000342: Could not obtain connection to query metadata: org.postgresql.util.PSQLException: FATAL: password authentication failed for user "username"
        at org.postgresql.core.v3.ConnectionFactoryImpl.doAuthentication(ConnectionFactoryImpl.java:646)
        at org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:180)
        at org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:235)
        at org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)
        at org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:223)
        at org.postgresql.Driver.makeConnection(Driver.java:400)
        at org.postgresql.Driver.connect(Driver.java:259)
        at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:226)
        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:535)
        at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:516)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:75)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1134)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

The application still runs without error with the command 'quarkus:dev'

**Comment by leonardowestphal** [comment]
@yrodiere 
I did other tests and identified that the 'SQL strings added more than once for: <schema_name>.<table_name>' error only occurs when using the persistence unit name equal to the database schema name. In my case the only difference between them was that the first letter was capitalized on PU name.

I believe that now you can identify the origin of the weird message and fix it, if necessary.

Thanks!

**Comment by yrodiere** [comment]
> Now It is clear that after version 2.1.4 it is mandatory to specify this property, but the error presented made it difficult to understand/discover.

The JDBC URL is mandatory *if* you want to use your own database. In dev mode and tests, the lack of JDBC URL should not cause any problem, as Quarkus will launch a container automatically if there's no JDBC URL.

> I did other tests and identified that the 'SQL strings added more than once for: <schema_name>.<table_name>' error only occurs when using the persistence unit name equal to the database schema name

This is... very, very strange. I definitely don't know how in the world this could lead to a problem, as the PU name and schema name are handled by very different parts of the codebase.

Maybe that was just a coincidence, and you changed something else in the config?

You really meant "PU name and schema name"? Not "PU name and datasource name"? I'm asking because I noticed a strange behavior when the PU name and datasource name are different: https://github.com/quarkusio/quarkus/issues/26507

> I believe that now you can identify the origin of the weird message and fix it, if necessary.

I'm sorry but no, I can't... And, believe me, I tried.

In this discussion, I'm counting three different error messages, multiple versions of Quarkus, and multiple sets of configuration properties, only one of which is complete and explicit.

I just can't put that together and build a reproducer: all I could get is the behavior I would expect (either it works, or I get a different error than you, and one that is expected considering the configuration).

If you want me to fix something, please provide a simple reproducer. If you want, you can start from my attempt:

```
git clone git@github.com:yrodiere/quarkus-quickstarts.git
git switch i25607-schemamanager-exception
cd hibernate-orm-quickstart
```

There's a script in that directory to launch postgres and initialize the databases: `launch-postgres.sh`.

**Comment by leonardowestphal** [comment]
I did some more tests to help you.

> This is... very, very strange. I definitely don't know how in the world this could lead to a problem, as the PU name and schema name are handled by very different parts of the codebase.
 Maybe that was just a coincidence, and you changed something else in the config?
You really meant "PU name and schema name"? Not "PU name and datasource name"? I'm asking because I noticed a strange behavior when the PU name and datasource name are different: https://github.com/quarkusio/quarkus/issues/26507

Yes it was just a coincidence. When I changed de PU name I also changed DS name.
You're right, the problem is the same as the issue [#26507 ](https://github.com/quarkusio/quarkus/issues/26507)

Hope this helps! Thanks for your attention.
